### PR TITLE
[IMP] website_sale: send an email to customers for their abandoned cart 

### PR DESCRIPTION
**Pull request by rga-odoo**
task:  https://www.odoo.com/web#id=29459&view_type=form&model=project.task&action=327&menu_id=
pad: https://pad.odoo.com/p/r.53374e7dcf15b2dfb833fc220d521d5f

### CODE REVIEW

**Comment by d-fence** [comment]
Not yet ready. I found that the email sent does not contain the access token in the url.

File: addons/website_sale/data/mail_template_data.xml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+            <field name="lang">${object.partner_id.lang}</field>
+            <field name="body_html" type="xml">
+              <div style="margin:auto;width:100%;">
+                <img src="/web/image/res.company/${user.company_id.id}/logo" style="height: auto; width: 80px; margin-top:30px;" alt="${user.company_id.name}"/>
+                <h1 style="color:#A9A9A9;">THERE'S SOMETHING IN YOUR CART.</h1>
+                <p>Would you like to complete your purchase?</p><br/><br/>
+                % if object.order_line:
+                  % for line in object.order_line:
+                    <hr/>
+                    <table width="100%">
```
Comment (rim-odoo) [inline addons/website_sale/data/mail_template_data.xml]: It means we will generate one table per order line. Can you just bump up the `<table>` tag outside the loop?

File: addons/website_sale/controllers/backend.py [hunk 2]
```diff
@@ -90,6 +90,7 @@ def fetch_dashboard_data(self, date_from, date_to):
                 ('order_line', '!=', 'False'),
                 ('partner_id', '!=', request.env.ref('base.public_partner').id),
                 ('date_order', '<=', fields.Datetime.to_string(datetime.now() - timedelta(hours=1))),
+                ('partner_id', '!=', request.env.ref('base.public_partner').id),
```
Comment (rim-odoo) [inline addons/website_sale/controllers/backend.py]: In your wizard hereunder, you add `('team_id.team_type', '=', 'website')` in the domain

File: addons/website_sale/controllers/backend.py [hunk 3]
```diff
@@ -90,6 +90,7 @@ def fetch_dashboard_data(self, date_from, date_to):
                 ('order_line', '!=', 'False'),
                 ('partner_id', '!=', request.env.ref('base.public_partner').id),
                 ('date_order', '<=', fields.Datetime.to_string(datetime.now() - timedelta(hours=1))),
```
Comment (rim-odoo) [inline addons/website_sale/controllers/backend.py]: `utcnow()` as you use it later?

File: addons/website_sale/models/mail_compose_message.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+import datetime
+from odoo import api, fields, models
+
+class MailComposeMessage(models.TransientModel):
+    _inherit = 'mail.compose.message'
+
+    @api.multi
+    def send_mail(self, auto_commit=False):
+        context = self._context
+        one_hour_before = datetime.datetime.utcnow() - datetime.timedelta(hours=1)
```
Comment (rim-odoo) [inline addons/website_sale/models/mail_compose_message.py]: [nitpicking] move this variable assignation under the `if` [/nitpicking]

File: addons/website_sale/models/sale_order.py [hunk 5]
```diff
@@ -2,6 +2,9 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 import logging
 import random
+import uuid
```
Comment (rim-odoo) [inline addons/website_sale/models/sale_order.py]: Unused import :)

File: addons/website_sale/models/sale_order.py [hunk 6]
```diff
@@ -40,6 +47,14 @@ def _compute_cart_info(self):
             order.only_services = all(l.product_id.type in ('service', 'digital') for l in order.website_order_line)
 
     @api.multi
+    @api.depends('team_id.team_type', 'date_order', 'order_line', 'state', 'partner_id')
+    def _compute_abandoned_cart(self):
+        time_constraint = fields.Datetime.to_string(datetime.now() - relativedelta(hours=1))
```
Comment (rim-odoo) [inline addons/website_sale/models/sale_order.py]: `utcnow()` ? :)

File: addons/website_sale/models/sale_order.py [hunk 7]
```diff
@@ -21,6 +24,10 @@ class SaleOrder(models.Model):
     )
     cart_quantity = fields.Integer(compute='_compute_cart_info', string='Cart Quantity')
     only_services = fields.Boolean(compute='_compute_cart_info', string='Only Services')
+    is_abandoned_cart = fields.Boolean(compute='_compute_abandoned_cart', string='Abandoned Cart')
```
Comment (rim-odoo) [inline addons/website_sale/models/sale_order.py]: If you store this field (easy) or make a search function (less easy) for it, you can simplify the code in your dashboard and in the wizard

File: addons/website_sale/views/templates.xml [hunk 8]
```diff
@@ -1799,4 +1799,21 @@
...
+        <xpath expr="//t[@t-call='website_sale.cart_lines']" position="before">
+            <div class="row">
+                <div t-if="proceed" class="mt8 bg-danger" style="padding: 15px;">
+                    <p>Your Previous Cart has been completed.</p>
+                    <p t-if="website_sale_order">Please proceed your current cart.</p>
+                </div>
+                <div t-if="token" class="mt8 bg-danger" style="padding: 15px;">
+                    <p>This is your Current cart.</p>
+                    <p><a t-attf-href="/shop/cart/#{token}?new=true">Click Here</a> if you want to restore your previous cart. Your Current cart will be replaced with your previous cart.</p>
+                    <p><a t-attf-href="/shop/cart/#{token}?merge=true">Click Here</a> if you want to merge your previous cart into current cart.</p>
```
Comment (rim-odoo) [inline addons/website_sale/views/templates.xml]: Wording is unclear: to be checked with @fgi-odoo

File: addons/website/static/src/less/website.backendDashboard.less [hunk 9]
```diff
@@ -137,6 +137,7 @@
                 height: 80px;
                 overflow: hidden;
                 text-overflow: ellipsis;
+                white-space: nowrap;
```
Comment (d-fence) [inline addons/website/static/src/less/website.backendDashboard.less]: It fixes that :
https://drive.google.com/file/d/0B0IF0OmAMb5OdXM0cnhGTnJMOXc/view

File: addons/website_sale/controllers/backend.py (hunk 2 above)
Comment (d-fence) [inline addons/website_sale/controllers/backend.py]: sale_order_domain at line 60 + 88 already include the team_id:
('team_id', 'in', request.env['crm.team'].search([('team_type', '=', 'website')]).ids)

File: addons/website_sale/data/mail_template_data.xml (hunk 1 above)
Comment (d-fence) [inline addons/website_sale/data/mail_template_data.xml]: Well, we need it if we want an Horizontal Rule between order line.

File: addons/website_sale/models/sale_order.py (hunk 7 above)
Comment (d-fence) [inline addons/website_sale/models/sale_order.py]: As we discussed IRL, as the field is based on the running time, we cannot store it.

File: addons/website_sale/controllers/main.py [hunk 10]
```diff
@@ -1003,3 +1003,29 @@ def country_infos(self, country, mode, **kw):
...
+            return request.render('website.404')
+        if order.state != 'draft': # check that cart has been procced or not.
+            response = self.cart()
+            response.qcontext.update({'proceed': 1})
+            return response
+        if post.get('new'):  # redirect to previous cart
+            request.session['sale_order_id'] = order.id
+            return request.redirect("/shop/cart")
+        if post.get('merge'): # merge previous cart into current cart
+            order.order_line.write({'order_id': request.session['sale_order_id']})
```
Comment (d-fence) [inline addons/website_sale/controllers/main.py]: As discussed IRL, I tried something but it didn't work as expected.

File: addons/website_sale/controllers/main.py [hunk 11]
```diff
@@ -997,3 +997,29 @@ def country_infos(self, country, mode, **kw):
...
+
+    @http.route("/shop/cart/<token>", type='http', auth="public", website=True)
+    def access_abandoned_cart(self, token=None, **post):
+        """
+        Token will be pass from revival email template and it will redirect to the selected cart.
+        """
+        order = request.env['sale.order'].sudo().search([('access_token', '=', token)], limit=1)
+        if not order:
+            return request.render('website.404')
+        if order.state != 'draft': # check that cart has been procced or not.
```
Comment (tde-banana-odoo) [inline addons/website_sale/controllers/main.py]: Don't know; should we specifically handle confirmed/ cancelled / already paid orders differently ?

File: addons/website_sale/controllers/main.py [hunk 12]
```diff
@@ -997,3 +997,29 @@ def country_infos(self, country, mode, **kw):
...
+            response = self.cart()
+            response.qcontext.update({'proceed': 1})
+            return response
+        if post.get('new'):  # redirect to previous cart
```
Comment (tde-banana-odoo) [inline addons/website_sale/controllers/main.py]: Well I would get it out of the **post and make it a real argument, as it is quite an important method parameter. I think it could be better named, because grepping 'new' in Odoo when you want to find where it is used and/or defined will give you a lot of results.

File: addons/website_sale/controllers/main.py [hunk 13]
```diff
@@ -997,3 +997,29 @@ def country_infos(self, country, mode, **kw):
...
+            request.session['sale_order_id'] = order.id
+            return request.redirect("/shop/cart")
+        if post.get('merge'): # merge previous cart into current cart
```
Comment (tde-banana-odoo) [inline addons/website_sale/controllers/main.py]: Same here actually. If post ocntains only two significant parameter, probably we should define them (and add them in the docstring ;) ) and get rid of the post.

File: addons/website_sale/controllers/main.py (hunk 10 above)
Comment (tde-banana-odoo) [inline addons/website_sale/controllers/main.py]: Good question. What is the expected result ?

File: addons/website_sale/controllers/main.py [hunk 14]
```diff
@@ -997,3 +997,29 @@ def country_infos(self, country, mode, **kw):
...
+            order.order_line.write({'order_id': request.session['sale_order_id']})
+            order.action_cancel()
+            return request.redirect("/shop/cart")
+        if request.session.get('sale_order_id') and order.id != request.session['sale_order_id']:
```
Comment (tde-banana-odoo) [inline addons/website_sale/controllers/main.py]: What is the purpose of that ?

File: addons/website_sale/models/mail_compose_message.py [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import datetime
+from odoo import api, fields, models
+
+class MailComposeMessage(models.TransientModel):
+    _inherit = 'mail.compose.message'
+
+    @api.multi
+    def send_mail(self, auto_commit=False):
```
Comment (tde-banana-odoo) [inline addons/website_sale/models/mail_compose_message.py]: Well what you are trying to do, if I understand well, is a bit tricky :) . I suppose you want to flag orders you just mailed that were abandoned carts. Several things here

1/ code optimization: we should avoid doing
```
records = self.search([])
for rec in records:
  rec.foo = bar
```
We should update records in batch and write on all of them at once
```
records = self.search([])
records.write({'foo': 'bar})
```
Note that this can be simplified into self.search([]).write({}), because you can pipe some recordsets methods.

2/ actually overriding the mail composer is not a good practice. You should look at mail.thread class and specifically at its message_post method. This involves inheritance, abstract classes, the chatter mecanism and a lot of funny stuff. I bet @rim-odoo will be glad to help you :p .

3/ mail_post_autofollow should be used only in some cases, but this is a bit complicated to explain here. First thing is probably to try to override message_post and maybe manually use message_subscribe (hahaaa, another key word).

A lot of key words in this comment. Be strong.

File: addons/website_sale/models/sale_order.py [hunk 16]
```diff
@@ -21,6 +23,10 @@ class SaleOrder(models.Model):
...
+    cart_recovery_email_sent = fields.Boolean(
+        default=False,
```
Comment (tde-banana-odoo) [inline addons/website_sale/models/sale_order.py]: As this is the default value of all boolean fields there is no need to repeat it. It simplifies the model declaration.

File: addons/website_sale/models/sale_order.py [hunk 17]
```diff
@@ -40,6 +46,14 @@ def _compute_cart_info(self):
             order.only_services = all(l.product_id.type in ('service', 'digital') for l in order.website_order_line)
 
     @api.multi
+    @api.depends('team_id.team_type', 'date_order', 'order_line', 'state', 'partner_id')
+    def _compute_abandoned_cart(self):
+        time_constraint = fields.Datetime.to_string(datetime.utcnow() - relativedelta(hours=1))
+        for order in self:
+            domain = order.date_order <= time_constraint and order.team_id.team_type == 'website' and order.state == 'draft' and order.partner_id.id != self.env.ref('base.public_partner').id and order.order_line
+            order.is_abandoned_cart = True if domain else False
```
Comment (tde-banana-odoo) [inline addons/website_sale/models/sale_order.py]: I bet you can simplify this line :) .

File: addons/website_sale/views/sale_order_views.xml [hunk 18]
```diff
@@ -7,10 +7,33 @@
...
                 <filter string="Confirmed Orders" name="order_confirmed" domain="[('state', 'in', ('sale', 'done'))]"/>
+                <separator/>
+                <filter string="From Website" name="from_website" domain="[('team_id.team_type', '=', 'website')]"/>
+                <separator/>
+                <filter string="Recovery Email to Send" name="recovery_email" domain="[('cart_recovery_email_sent', '=', False)]" invisible="1"/>
             </search>
         </field>
     </record>
 
+    <record id="view_order_form_inherit_website_sale" model="ir.ui.view">
```
Comment (tde-banana-odoo) [inline addons/website_sale/views/sale_order_views.xml]: JustNamingThings: we have some guidelines on how to write xml_ids. For example for a view
```
<model>_view_<view_type>_<suffix_for_details>
```
It could be sale_order_view_form_inherit_cart_recovery for example.

File: addons/website_sale/controllers/main.py [hunk 19]
```diff
@@ -997,3 +997,29 @@ def country_infos(self, country, mode, **kw):
             states=[(st.id, st.name, st.code) for st in country.get_website_sale_states(mode=mode)],
             phone_code=country.phone_code
         )
+
+    @http.route("/shop/cart/<token>", type='http', auth="public", website=True)
+    def access_abandoned_cart(self, token=None, **post):
+        """
+        Token will be pass from revival email template and it will redirect to the selected cart.
+        """
+        order = request.env['sale.order'].sudo().search([('access_token', '=', token)], limit=1)
```
Comment (d-fence) [inline addons/website_sale/controllers/main.py]: You are redirected to your current cart content:
[cart route is '/shop/cart/'](https://github.com/odoo-dev/odoo/blob/2f791e81927ad9f9eed1856554cc5404af08d8b9/addons/website_sale/controllers/main.py#L337)

File: addons/website_sale/controllers/main.py (hunk 11 above)
Comment (d-fence) [inline addons/website_sale/controllers/main.py]: Yes, because abandoned carts are 'draft'. 'canceled' could made sense if the seller canceled the cart but the customer really wants to get it back ?!
'already paid', what could be the purpose of getting back an already paid cart ? I think that it could be confusing for the customer to see a cart coming back from the limbo.

File: addons/website_sale/controllers/main.py (hunk 10 above)
Comment (d-fence) [inline addons/website_sale/controllers/main.py]: As the customer already filled (at least with essential informations) the partner form in his previous cart, it would be great to get the form pre-filled when he will eventually proceed the cart.

File: addons/website_sale/controllers/main.py (hunk 14 above)
Comment (d-fence) [inline addons/website_sale/controllers/main.py]: It's the case where the customer already filled the filled the "billing address form" with his current cart and then he followed the link to get back his abandoned cart.

File: addons/website_sale/controllers/backend.py [hunk 20]
```diff
@@ -89,7 +89,7 @@ def fetch_dashboard_data(self, date_from, date_to):
                 ('state', '=', 'draft'),
                 ('order_line', '!=', 'False'),
                 ('partner_id', '!=', request.env.ref('base.public_partner').id),
-                ('date_order', '<=', fields.Datetime.to_string(datetime.now() - timedelta(hours=1))),
+                ('date_order', '<=', fields.Datetime.to_string(datetime.utcnow() - timedelta(hours=1))),
```
Comment (d-fence) [inline addons/website_sale/controllers/backend.py]: Problem is that it returns a strings which need to be re-converted to use in timedelta calculations.

File: addons/website_sale/models/mail_compose_message.py (hunk 15 above)
Comment (d-fence) [inline addons/website_sale/models/mail_compose_message.py]: A mail wizard is the expected behavior

File: addons/website_sale/models/sale_order.py (hunk 17 above)
Comment (d-fence) [inline addons/website_sale/models/sale_order.py]: Bof

File: addons/website_sale/controllers/main.py [hunk 21]
```diff
@@ -334,20 +334,45 @@ def pricelist(self, promo, **post):
         return request.redirect(redirect)
 
     @http.route(['/shop/cart'], type='http', auth="public", website=True)
-    def cart(self, **post):
+    def cart(self, abd_token=None, revive='', **post):
```
Comment (d-fence) [inline addons/website_sale/controllers/main.py]: Rename abd_toke in access_token

File: addons/website_sale/models/mail_compose_message.py [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+        context = self._context
+        if context.get('default_model') == 'sale.order':
```
Comment (d-fence) [inline addons/website_sale/models/mail_compose_message.py]: Use self

File: addons/website_sale/models/mail_compose_message.py [hunk 23]
```diff
@@ -0,0 +1,26 @@
...
+                ('id', 'in', context.get('active_ids')),
+                ('cart_recovery_email_sent', '!=', True),
+                ('state', '=', 'draft'),
+                ('partner_id.id', '!=', self.env.ref('base.public_partner').id),
+                ('order_line', '!=', False),
+                ('team_id.team_type', '=', 'website'),
+                ('date_order', '<', fields.Datetime.to_string(one_hour_before))]
+            abandonned_orders = self.env['sale.order'].search(abandoned_cart_domain)
+            abandonned_orders.write({'cart_recovery_email_sent': True})
+            self = self.with_context(mail_post_autofollow=True)
```
Comment (tde-banana-odoo) [inline addons/website_sale/models/mail_compose_message.py]: Is it necessary to add the autofollow mechanism ? If you do it in batch it means you will be follower of all sale orders.

File: addons/website_sale/models/sale_order.py (hunk 17 above)
Comment (tde-banana-odoo) [inline addons/website_sale/models/sale_order.py]: Yes you can ! Like order.is_abandoned_cart = bool(domain) .

File: addons/website_sale/models/sale_order.py [hunk 24]
```diff
@@ -174,6 +187,28 @@ def _cart_accessories(self):
...
 
+    @api.multi
+    def action_recovery_email_send(self):
+        compose_form_id = self.env.ref('mail.email_compose_message_wizard_form').id
+        template_id = self.env.ref('website_sale.recovery_mail').id
+        ctx = {
+            'default_composition_mode': 'mass_mail' if len(self) > 1 else 'comment',
+            'default_res_id': self.ids[0],
+            'default_model': 'sale.order',
+            'default_use_template': bool(template_id),
```
Comment (tde-banana-odoo) [inline addons/website_sale/models/sale_order.py]: This is strange. If the template does not exists you method will crash at template_id = self.env.ref('website_sale.recovery_mail').id because ref either crashes either returns None when the xml_id is not present anymore.

File: addons/website_sale/views/crm_team_views.xml [hunk 25]
```diff
@@ -16,8 +16,8 @@
                             <div>
                                 <a name="get_abandoned_carts" type="object">
                                     <field name="abandoned_carts_count"/>
-                                    <t t-if="record.abandoned_carts_count.raw_value != 1">Abandoned Carts</t>
-                                    <t t-else="">Abandoned Cart</t>
+                                    <t t-if="record.abandoned_carts_count.raw_value != 1">Abandoned Carts to Recover</t>
```
Comment (tde-banana-odoo) [inline addons/website_sale/views/crm_team_views.xml]: Actually I would remove this t-if t-else. In Odoo we generally use plural to avoid having if / else everywhere. I would therefore always display Abandoned Carts to Recover.

File: addons/website_sale/views/sale_order_views.xml [hunk 26]
```diff
@@ -7,10 +7,33 @@
...
+                <separator/>
+                <filter string="From Website" name="from_website" domain="[('team_id.team_type', '=', 'website')]"/>
+                <separator/>
+                <filter string="Recovery Email to Send" name="recovery_email" domain="[('cart_recovery_email_sent', '=', False)]" />
             </search>
         </field>
     </record>
 
+    <record id="sale_order_view_form_cart_recovery" model="ir.ui.view">
+        <field name="name">sale.order.form.acquirer</field>
```
Comment (tde-banana-odoo) [inline addons/website_sale/views/sale_order_views.xml]: Why is the name a copy-and-paste from acquirer addon :p ?

File: addons/website_sale/views/sale_order_views.xml [hunk 27]
```diff
@@ -7,10 +7,33 @@
...
+        <field name="model">sale.order</field>
+        <field name="inherit_id" ref="sale.view_order_form"/>
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='team_id']" position="after">
+                <field name="is_abandoned_cart" invisible="1"/>
+                <field name="cart_recovery_email_sent" invisible="1"/>
+            </xpath>
+            <xpath expr="//button[@name='action_quotation_send' and @states='draft']" position="attributes">
```
Comment (tde-banana-odoo) [inline addons/website_sale/views/sale_order_views.xml]: I would change the view_order_form view to remove the "states=draft" and use a real invisible attribute. Then it will be easier to override here because it is based on some implementation details of the web client.

File: addons/website_sale/views/sale_order_views.xml [hunk 28]
```diff
@@ -53,6 +76,26 @@
         <field name="help">You don't have any order from the website.</field>
     </record>
 
+    <!-- Server action to send multiple recovery email-->
+    <record id="send_recovery_email_action_server" model="ir.actions.server">
+        <field name="name">Send a Recovery Email</field>
```
Comment (tde-banana-odoo) [inline addons/website_sale/views/sale_order_views.xml]: As for mail templates we put the main business model in the server action name. That way when viewing the server action list they are ordered by name and therefore by main model. Here it would be Sale: Send Cart Recovery Email or something like that.

File: addons/website_sale/views/sale_order_views.xml [hunk 29]
```diff
@@ -53,6 +76,26 @@
...
+        <field name="model_id" ref="model_sale_order"/>
+        <field name="state">code</field>
+        <field name="code">
+            if records:
+                action = records.action_recovery_email_send()
+        </field>
+    </record>
+
+    <record id="action_send_recovery_email" model="ir.values">
+        <field name="name">Send a Recovery Email</field>
```
Comment (tde-banana-odoo) [inline addons/website_sale/views/sale_order_views.xml]: Same as server action name.

File: addons/website_sale/views/sale_order_views.xml [hunk 30]
```diff
@@ -66,7 +109,7 @@
     <menuitem id="website.menu_website_global_configuration" name="Configuration"
         parent="website.menu_website_configuration" sequence="100"
         groups="base.group_system,sales_team.group_sale_manager"/>
-  
+
```
Comment (tde-banana-odoo) [inline addons/website_sale/views/sale_order_views.xml]: Unnecessary diff

File: addons/website_sale/views/templates.xml [hunk 31]
```diff
@@ -1764,4 +1764,21 @@
           </div>
         </t>
     </template>
+
+    <!-- When open cart from abandoned email template and already existing diffrent cart then link in session, Ask for replace/merge the existing cart with the new cart -->
+    <template id="abandoned_cart_restore" inherit_id="website_sale.cart" name="abandoned cart restore">
```
Comment (tde-banana-odoo) [inline addons/website_sale/views/templates.xml]: Why is it in a separate template and not directly in website_sale.cart template ?

File: addons/website_sale/views/sale_order_views.xml [hunk 32]
```diff
@@ -53,6 +76,26 @@
         <field name="help">You don't have any order from the website.</field>
     </record>
 
+    <!-- Server action to send multiple recovery email-->
+    <record id="send_recovery_email_action_server" model="ir.actions.server">
```
Comment (tde-banana-odoo) [inline addons/website_sale/views/sale_order_views.xml]: I would rename the xml_id to something like ir_actions_server_sale_cart_recovery, for the same reasong as already mentioned before :) . I know, naming conventions ...

File: addons/website_sale/data/mail_template_data.xml [hunk 33]
```diff
@@ -0,0 +1,40 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <data noupdate="1">
+        <!--Email template -->
+        <record id="recovery_mail" model="mail.template">
+            <field name="name">Abandoned Cart - Send a Recovery Email</field>
+            <field name="email_from">${(object.user_id.email or '')|safe}</field>
```
Comment (d-fence) [inline addons/website_sale/data/mail_template_data.xml]: No because I think that a user created from the website sale shop does not have a parent_id by default and the email field is mandatory when proceeding the cart.

File: addons/website_sale/views/sale_order_views.xml (hunk 26 above)
Comment (d-fence) [inline addons/website_sale/views/sale_order_views.xml]: I dunno :-)

File: addons/website_sale/models/sale_order.py (hunk 24 above)
Comment (d-fence) [inline addons/website_sale/models/sale_order.py]: Changed to raise an exception. Not sure that it's the correct way to do it.

File: addons/website_sale/views/templates.xml (hunk 31 above)
Comment (d-fence) [inline addons/website_sale/views/templates.xml]: Probably because this whole template file is constructed the same way. Personally I find it easier/less messy when it's separated.
